java-checks/src/main/java/org/sonar/java/checks/PatternMatchUsingIfCheck.java

```diff
@@ -0,0 +1,280 @@
+    Preconditions.checkArgument(!cases.isEmpty());
+    var iter = cases.iterator();
+    var scrutinee = iter.next().scrutinee();
+    while (iter.hasNext()) {
+      if (!iter.next().scrutinee().equals(scrutinee)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static @Nullable List<Case> extractCasesFromIfSequence(IfStatementTree topLevelIfStat) {
+    var cases = new LinkedList<Case>();
+    StatementTree stat;
+    for (stat = topLevelIfStat; stat instanceof IfStatementTree ifStat; stat = ifStat.elseStatement()) {
+      var caze = convertToCase(ifStat.condition(), ifStat.thenStatement());
+      if (caze == null) {
+        return null;
+      }
+      cases.add(caze);
+    }
+    if (stat != null) {
+      cases.add(new DefaultCase(cases.getLast().scrutinee(), stat));
+    }
+    return cases;
+  }
+
+  private static @Nullable Case convertToCase(ExpressionTree condition, StatementTree body) {
+    var leftmost = findLeftmostInConjunction(condition);
+    var guards = new LinkedList<ExpressionTree>();
+    populateGuardsList(condition, guards);
+    if (leftmost instanceof PatternInstanceOfTree patInstOf && patInstOf.pattern() != null
+      && patInstOf.expression() instanceof IdentifierTree idTree) {
+      return new PatternMatchCase(idTree.name(), patInstOf.pattern(), guards, body);
+    } else if ((leftmost.kind() == Tree.Kind.CONDITIONAL_OR || leftmost.kind() == Tree.Kind.EQUAL_TO) && guards.isEmpty()) {
+      return buildEqualityCase(leftmost, body);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Transforms expressions of the form  a == 0 || a == 1 || ...  into an EqualityCase
+   */
+  private static @Nullable EqualityCase buildEqualityCase(ExpressionTree expr, StatementTree body) {
+    var constantsList = new LinkedList<ExpressionTree>();
+    String scrutinee = null;
+    while (expr.kind() == Tree.Kind.CONDITIONAL_OR) {
+      var binary = (BinaryExpressionTree) expr;
+      var varAndCst = extractVarAndConstFromEqualityCheck(binary.rightOperand());
+      if (varAndCst == null) {
+        return null;
+      } else if (scrutinee == null) {
+        scrutinee = varAndCst.a;
+      } else if (!varAndCst.a.equals(scrutinee)) {
+        return null;
+      }
+      constantsList.addFirst(varAndCst.b);
+      expr = binary.leftOperand();
+    }
+    var varAndCst = extractVarAndConstFromEqualityCheck(expr);
+    if (varAndCst == null || (scrutinee != null && !varAndCst.a.equals(scrutinee))) {
+      return null;
+    }
+    constantsList.addFirst(varAndCst.b);
+    return new EqualityCase(scrutinee == null ? varAndCst.a : scrutinee, constantsList, body);
+  }
+
+  private static @Nullable Pair<String, ExpressionTree> extractVarAndConstFromEqualityCheck(ExpressionTree expr) {
+    if (expr.kind() == Tree.Kind.EQUAL_TO) {
+      var binary = (BinaryExpressionTree) expr;
+      if (binary.leftOperand() instanceof IdentifierTree idTree && isPossibleConstantForCase(binary.rightOperand())) {
```

Comment:
Here you are taking for granted that you will only find conditions like ` x == 2 || x ==3 `, and the rule will not raise when the two elements of the binary expression are in the opposite order, like ` x == 2 || 3 == x `

This case is a FN
```
  void falseNegative(int x) {
    if (x == 1 || 3 == x) {
      System.out.println("one");
    } else if (x == 2) {
      System.out.println("two");
    }
  }
```